documentation/proposals/Proposal - Generation of Library Sources and PInvoke Mechanisms.md

```diff
@@ -6,283 +6,268 @@ Proposal design for a platform invoke (P/Invoke) mechanism for Silk.NET 3.0.
 
-`UseStaticLibrary` **MUST** be able to be specified on either a function or type.
+This naturally makes us entirely dependent on an external dependency, but I propose we work with Tanner Gooding (the ClangSharp project lead) as much as possible to add the functionality we need in the least breaking way possible, and in a way that satisfies both us and ClangSharp. All designs for such modifications will be formalized in the ClangSharp repo. Should we fail to do this, we'll maintain a fork so we can still benefit from improvements made upstream, while giving ourselves the freedom to add the functionality we need.
 
-Consider the following example:
-```cs
-#if __IOS__
-[UseStaticLibrary]
-#endif
-[UseDynamicLibrary("glfw3.dll")]
-public partial class Glfw
-{
-    public partial void glfwInit();
-}
-```
+Tanner has already stated that they're happy to work with us to get Silk.NET using ClangSharp, even saying they're happy to add a CI test stage into the ClangSharp repo to ensure no incoming changes break Silk.NET's generation process.
 
-The Emitter **MUST** only generate code if all preprocessor directives guarding the `UseStaticLibrary` attribute evaluate to true. If the attribute is defined on both the function and the containing type, the preprocessor directives surrounding the function's attribute **MUST** be used instead of the preprocessor directives surrounding the type.
+# Implicit Cast-Based Overloading
 
-Note to the reader: Given preprocessor directives are processed at parse time in Roslyn, both of those last requirements are basically benign.
+The pivotal proposal around which the SilkX experiment revolves is solving the excessive overload problem by having no overloads at all. Namely, Silk.NET.Core will have a set of "pointer-like" types that can be implicitly casted to by a multitude of different types. This gives the user the same flexibility as having all different permutations of overloaded parameters generated without having to do so, and thereby alleviates the requirement of having overloads at all.
 
-#### Custom Call Style: Procedure Address Expressions
+`Ptr` and `Mut` `ref struct`s **shall** be defined wrapping a `ref readonly` and `ref` respectively. These types **shall** have generic and non-generic variants. This shall be repeated up to 3 pointer dimensions, taking into account C's mutability rules in each instantiation. This means that there will be 14 top-level generic structs and 14 top-level non-generic structs. These **should** be auto-generated to ensure they remain maintainable. IL weaving **may** be required for the higher dimensions.
 
-Custom code may be used as a call style by providing a pointer to SilkTouch using a Procedure Address Expression. This is useful in scenarios such as COM interop.
+Unless deemed inappropriate, inapplicable, and/or infeasible by the Silk.NET team (the team reserved the right to do so without Working Group approval), the types **should** have the following characteristics:
```

Comment:
In the case of a generic pointer type, implicitly convertible to its non-generic counterpart?

Reply:
This should be easy enough as we would just be passing the InteriorRef as the Ref for the non-generic consturctor.